### DISCUSSION TITLE
DirectGraph Layout not starting from origin(0,0) in autoResize:true ,borderless:true paperScroller

### DISCUSSION THREAD

**Discussion by jibinjoy8897** [clientIO/joint#2534] [Q&A] [answered]
### Introduction

My current paper scroller settings
const scroller = new ui.PaperScroller({
      paper,
      cursor: 'grab',
      inertia: { friction: 0.8 },
      borderless: true,
      autoResizePaper: true,
      contentOptions: function () {
        return {
          useModelGeometry: true,
          allowNewOrigin: 'positive',
        };
      },
    });

Directed Graph 

layout.DirectedGraph.layout(graph, {
      graphlib: dagre.graphlib,
      dagre: dagre,
      setVertices: true,
      rankDir: 'TB',
      marginX: 100,
      marginY: 100,
    });

In this scenario,my diagram starts from a large offset from top and left (0,0).The customer wants it starting almost near 0,0(origin).Please suggest solution for the same.

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS+

**Comment by kumilingus** [comment]
Set `marginX: 0` and `marginY: 0`.

**Reply by jibinjoy8897** [reply]
Tried this solution.No change happened.When inspected g tag inside svg tag is having offset from a large distance from 0,0

**Reply by kumilingus** [reply]
Are you referring to the viewport `<g/>` transformation or the paper scroller `padding` this time?

**Reply by jibinjoy8897** [reply]
g tag.Paper scroller paper padding is fixed.Svg tag starts from 0,0 and is working perfectly

**Reply by kumilingus** [reply]
You can not use the `borderless` option if you don't want the paper to extend beyond the `0,0` origin.

**Reply by jibinjoy8897** [reply]
I do not want to paper to extend beyond (0,0).I just want my directedLayout to start near 0,0 without offset.Can you suggest a way to avoid the offset between svg and g tag?

**Reply by kumilingus** [reply]
I just told you.

> You can not use the borderless option if you don't want the paper to extend beyond the 0,0 origin.

---

Here's an example:

```js
const paperScroller = new ui.PaperScroller({
    paper,
    baseWidth: 1000,
    baseHeight: 1000,
    autoResizePaper: true,
    contentOptions: {
        // @ts-ignore to be fixed in the next version
        allowNewOrigin: false,
    }
});

DirectedGraph.layout(graph, {
    marginX: 0,
    marginY: 0
});
```

**Comment by jibinjoy8897** [comment]
![OriginIssue](https://github.com/clientIO/joint/assets/56633789/b68d6dcb-08fd-4812-9970-80db6abb9282)
I have implemented your code given above and the image shows the result.The svg is perfect and starts from 0,0(confirmed it by elements not able to move to the negative coordinates ie the grey area (paper scroller)outside the svg tag.

Issue:
In the image the scrolls have more space to the left and top to move.
How to align this paper(svg) to the paperScroller(grey area) with 0,0 (origin) and so that the scroll starts from the extreme top and left indicating the left most part of the paper is reached.

**Reply by kumilingus** [reply]
You can set `padding` to `0`, but even that won't guarantee that the origin of `0,0` will be in the top left corner when `scrollLeft` and `scrollTop` are `0` (especially when you zoom out).

--- 

You can try the following code. Custom PaperScoller, which doesn't add any padding (but it has its drawbacks, especially when zooming).

```js
class MyPaperScroller extends joint.ui.PaperScroller {
    computeRequiredPadding() {
        return {
            left: 0,
            right: 0,
            top: 0,
            bottom: 0
        }
    }
}

const paperScroller = new MyPaperScroller({
    paper: paper,
    autoResizePaper: true,
    scrollWhileDragging: true,
    cursor: 'grab',
    padding: 0,
    contentOptions: {
        allowNewOrigin: false
    }
});
```

**Reply by jibinjoy8897** [reply]
I have added the code given.Works ok but the paper is not autoResizing when we drag outside the positive initial limits but the shrinking of paper is happening when we group tables to left.Once paper shrinks to a lower size it doesnot change back.

**Reply by kumilingus** [reply]
Unfortunately, I had no idea how to reproduce this. What "grouping tables" means.

**Reply by jibinjoy8897** [reply]
Grouping tables means dragging the tables from the right part of the extended canvas to the left ,just for trying to make the paper shrink in size.Please suggest a way to handle scroll jump in this scenario.

**Reply by jibinjoy8897** [reply]
![ScrollJumpIssue](https://github.com/clientIO/joint/assets/56633789/a6fc4c95-36eb-4663-a872-1304dd07f608)
Please see the gif after your code suggestion.Scroll Jumping can we seen when paper is shrinking.

**Reply by kumilingus** [reply, marked as answer]
Perhaps you can redefine the `computeRequiredPadding()` as follows:

```js
class MyPaperScroller extends joint.ui.PaperScroller {

    computeRequiredPadding(...args) {
        const { right, bottom } = super.computeRequiredPadding(...args);
        return {
            left: 0,
            top: 0,
            right,
            bottom
        }
    }
}
```

**Reply by jibinjoy8897** [reply]
![ScrollDownIssue](https://github.com/clientIO/joint/assets/56633789/c9e5e7e9-e140-4d36-b3b0-cc0df167dcc4)
Scroll Jumping on the right is fixed after code change.But paper is not resizing properly in the bottom as shown in Gif

**Reply by kumilingus** [reply]
I am not able to reproduce this. Could you provide steps how to reproduce this in the Kitchen Sink application?

**Reply by jibinjoy8897** [reply]
Add restrictTranslate: true, in paper Options.The paper is not autoResizing after that

**Reply by kumilingus** [reply]
Yes, that is expected. If you set `restrictTranslate` to `true`, it will not allow elements to cross the paper boundary and therefore the paper size will not change.

You need to set the `restrictTranslate` as a function and only restrict the minus coordinates.

**Reply by jibinjoy8897** [reply]
can you give example for restrictTranslate as function for only minus coordinates?

**Reply by kumilingus** [reply]
Please refer to our [documentation](https://resources.jointjs.com/docs/jointjs/v4.0/joint.html#dia.Paper.prototype.options.restrictTranslate).
